core/server/common/src/main/java/alluxio/master/journal/AsyncJournalWriter.java

```diff
@@ -99,49 +155,177 @@ public long appendEntry(JournalEntry entry) {
+    mStopFlushing = false;
+    mFlushThread.start();
+  }
+
+  /**
+   * A dedicated thread that goes over outstanding queue items and writes/flushes them. Other
+   * threads can track progress by submitting tickets via ::flush() call.
+   */
+  private void doFlush() {
+    // Runs the loop until ::stop() is called.
+    while (!mStopFlushing) {
+
+      /**
+       * Stand still unless;
+       * - queue has items
+       * - permit is given by:
+       *   - clients
+       *   -::stop()
+       */
+      while (mQueue.isEmpty() && !mStopFlushing) {
+        try {
+          // Wait for permit up to batch timeout.
+          // PS: We don't wait for permit indefinitely in order to process
+          // queued entries proactively.
+          if (mFlushSemaphore.tryAcquire(mFlushBatchTimeNs, TimeUnit.NANOSECONDS)) {
+            break;
+          }
+        } catch (InterruptedException ie) {
+          break;
+        }
       }
-      long writeCounter = mWriteCounter.get();
-      while (targetCounter > writeCounter) {
-        for (;;) {
+
+      try {
+        long startTime = System.nanoTime();
+
+        // Write pending entries to journal.
+        while (!mQueue.isEmpty()) {
           // Get, but do not remove, the head entry.
           JournalEntry entry = mQueue.peek();
           if (entry == null) {
-            // No more entries in the queue. Break out of the infinite for-loop.
+            // No more entries in the queue. Break write session.
             break;
           }
           mJournalWriter.write(entry);
           // Remove the head entry, after the entry was successfully written.
           mQueue.poll();
-          writeCounter = mWriteCounter.incrementAndGet();
+          mWriteCounter++;
 
-          if (writeCounter >= targetCounter) {
-            if ((System.nanoTime() - startTime) >= mFlushBatchTimeNs) {
-              // This thread has been writing to the journal for enough time. Break out of the
-              // infinite for-loop.
-              break;
+          if (((System.nanoTime() - startTime) >= mFlushBatchTimeNs) && !mStopFlushing) {
+            // This thread has been writing to the journal for enough time. Break out of the
+            // infinite while-loop.
+            break;
+          }
+        }
+
+        // Either written new entries or previous flush had been failed.
+        if (mFlushCounter.get() < mWriteCounter) {
+          mJournalWriter.flush();
+          mFlushCounter.set(mWriteCounter);
+        }
+
+        // Notify tickets that have been served to wake up.
+        synchronized (mTicketList) {
+          ListIterator<FlushTicket> ticketIterator = mTicketList.listIterator();
+          while (ticketIterator.hasNext()) {
+            FlushTicket ticket = ticketIterator.next();
+            if (ticket.getTargetCounter() <= mFlushCounter.get()) {
+              ticket.setCompleted();
+              ticketIterator.remove();
```

Comment:
is `remove` efficient for an `ArrayList`?

Reply:
True. I changed it to linked list in order to favor flush performance.